src/envoy/http/authn/authn_utils.cc

```diff
@@ -87,21 +72,32 @@ bool AuthnUtils::GetJWTPayloadFromHeaders(
   }
   ::google::protobuf::Map< ::std::string, ::std::string>* claims =
       payload->mutable_claims();
+  Envoy::Json::ObjectSharedPtr json_obj;
   try {
-    auto json_obj = Json::Factory::loadFromString(payload_str);
+    json_obj = Json::Factory::loadFromString(payload_str);
     ENVOY_LOG(debug, "{}: json object is {}", __FUNCTION__,
               json_obj->asJsonString());
-    json_obj->iterate(
-        [payload](const std::string& key, const Json::Object& obj) -> bool {
-          return JsonIterCallback::JsonIteratorCallback(key, obj, payload);
-        });
-
   } catch (...) {
-    ENVOY_LOG(error, "Invalid {} header, invalid json: {}",
-              jwt_payload_key.get(), payload_str);
     return false;
   }
 
+  // Extract claims
+  json_obj->iterate(
+      [payload](const std::string& key, const Json::Object& obj) -> bool {
+        ::google::protobuf::Map< ::std::string, ::std::string>* claims =
+            payload->mutable_claims();
+        // In current implementation, only string objects are extracted into
+        // claims. If call obj.asJsonString(), will get "panic: not reached"
+        // from json_loader.cc.
+        try {
+          // Try as string, will throw execption if object type is not string.
+          (*claims)[key] = obj.asString();
+        } catch (Json::Exception& e) {
+        }
+        return true;
+      });
+  // Extract audience
+  ExtractJwtAudience(*json_obj, payload);
   if (payload->claims().empty()) {
     ENVOY_LOG(error, "{}: there is no JWT claims.", __func__);
```

Comment:
Is this log misleading if there are JWT claims but we fail to parse them as string and throw an exception at line 94?
Would it be necessary to distinguish errors between empty claim and failure of parsing claim?

Reply:
Done: the log has been removed.

Reply:
Good catch. I think we don't even need this check. 

If JWT filter output empty payload as its authentication result (it should not, but assume it does), then we should still consume the payload normally. The authN output will have no claims (nor principal, audiences etc), but it would be authZ to decide to reject request or not.

Reply:
Done.